### ISSUE TITLE
[Q] sync tensorboard step issue

### ISSUE THREAD

**Issue by jingxixu** [wandb/wandb#2369]
Hi,

Thanks for this amazing piece of program! 

I am switching from pytorch + tensorboard to wandb. I use the argument sync_tensorboard = True in wandb.init() and then create the tensorboard summary writer. When I add a scalar to tensorboard, I can customize the global_step argument so that each metric has a different axis (sometimes being epochs, episodes, training steps, etc.). I noticed that wandb has its own step counter and I also noticed that I can change the x-axis in the wandb online GUI. 

The problem is, when I am trying to log a metric, I pass the tensorboard writer as an object to another class defined in a different file, but this metric has a different global_step from the other logged metrics, and then when I check the online panel, the x-axis is messed up. I have no idea what that number is. When I switch to the default wandb "Step", it displays a weird number, and when I switch to "global_step" (which I believe should be the step I passed into my tensorboard writer), it becomes the same step as the other metrics, logged in the main training file. Neither of them is what I passed to the tensorboard writer.

I understand that my problem is quite hard to understand. So a general question is, what is the recommended workflow for switching from tensorboard to wandb? **Especially I am using tensorboard to log a lot of different metrics with different customized x axis**? I notice that the online doc mentioned wandb.tensorboard.pacth, what exactly is this? I did not find any online doc for this function and how is it different from setting sync_tensorboard = True in wandb.init()?

Or, if I am gonna abandon tensorboard completely, what are the most convenient ways to log different metrics, with different **CUSTOMIZED** x axis?

Thanks a lot!

Sincerely

**Comment by vanpelt** [comment]
Hey @jingxixu,

With tensorboard we log custom x-axis as "global_step" for the global namespace.  If you are logging to multiple tfevent files in different folders, we'll also log a step for those namespaced metrics i.e. "val/global_step" which you can change your charts to use.  Can you share a link to one of your runs so I can verify that namespaced metrics are getting logged?

**Comment by jingxixu** [comment]
Hi,

I am not logging to multiple tfevent files in different folders. How do I share a link to one of my runs?

Also, any comments about

> So a general question is, what is the recommended workflow for switching from tensorboard to wandb? Especially I am using tensorboard to log a lot of different metrics with different customized x axis? I notice that the online doc mentioned wandb.tensorboard.pacth, what exactly is this? I did not find any online doc for this function and how is it different from setting sync_tensorboard = True in wandb.init()?

> Or, if I am gonna abandon tensorboard completely, what are the most convenient ways to log different metrics, with different CUSTOMIZED x axis?

**Comment by jingxixu** [comment]
Just to clarify the metric got messed up. I pass the tensorboard writer as an argument to a member function of an object. The class of that object is defined in a different file.

**Comment by vanpelt** [comment]
Hey @jingxixu, tensorboard only supports logging a single "global_step" per metric.  We should capture this step value in the "global_step" x-axis.  `sync_tensorboard=True` just calls `wandb.tensorboard.patch` under the hood.  The only reason to call `wandb.tensorboard.patch` would be if you're logging to multiple tfevent files, you can specify the root log directory with the `root_logdir` argument of `wandb.tensorboard.patch`.

When logging metrics directly with wandb that have separate x-axis you can just log them as seperate metrics.  I.E.

```
wandb.log({"acc": X, "batch": Y})
```

Then in the UI you could choose "batch" as your x-axis for "acc".  If you want to have metrics automatically use a custom x-axis you can define them ahead of time by calling the following after you call wandb.init

```
wandb.run.define_metric("acc", step_metric="batch", summary="max")
```

That should make the charts use "batch" as the x-axis and also store the maximum value of "acc" seen during a run.  In general it's best practice to log all metrics that share the same step in the same call to `wandb.log`, so if you had other metrics that should also be logged at "batch" you can add them to your `wandb.log` call, or if you need to call them in seperate places in your code you can do the following:

```
wandb.log({"acc": acc, "batch": batch}, commit=False)
# Some other code, or another file that gets executed in the same batch
wandb.log({"loss": loss})
# The above is equivalent to calling
wandb.log({"acc": loss, "batch": batch, "loss": loss})
```

Let me know if you have additional questions.

**Comment by jingxixu** [comment]
Hi, this is great thanks!

I have another question about the log function, as described here: https://docs.wandb.ai/ref/python/log.

In the description of the `step` argument, it says

> This persists any non-committed earlier steps but defaults to not committing the specified step

What does this mean exactly? It defaults to not committing the specified step? I understand that each time I call `wandb.log` the `global_step` will be incremented automatically by default, is this `step` argument overwriting the `global_step`?

**Comment by vanpelt** [comment]
Sorry that's unclear.  "global_step" is a tensorboard concept and automatically added as a metric that charts can use as a custom x-axis.  The "step" keyword argument of `wandb.log` is overloaded.  It's both the number of times `wandb.log` is called (when step is not specified), and an only incrementing default x-axis value that "commits" the accumulated metrics of previous calls to `wandb.log` with the same step when it increases I.E.

```
wandb.log({"acc": 0.8}, step=1000)
wandb.log({"loss": 0.1}, step=1000)
wandb.log({"acc": 0.9}, step=1100)
```

is equivalent to (when "global_step" is chosen as an x-axis):

```
wandb.log({"acc": 0.8, "global_step": 1000}, commit=False)
wandb.log({"loss": 0.1})
wandb.log({"acc": 0.9, "global_step": 1100}, commit=False)
```

**Comment by jingxixu** [comment]
Just to make sure that I understand this correctly. If I have something like `wandb.log({"acc": 0.8}, step=1000)`, the data `{"acc": 0.8}` won't be committed until an actual commit happens. When `wandb.log({"acc": 0.9}, step=1100)` is called, all logs with step < 1100 will be committed but not this particular log itself.

**Comment by jingxixu** [comment]
I think now I understand the `wandb.log` better. Let me reiterate my original error because I still think there might still be a bug in the way wandb does tensorboard syncing.

Basically, I have two files. 
- `mina.py`: the main file
-  `A.py`: where class A is defined.

In the `main.py` file

```
wandb.init(sync_tensorboard=True, ...)
writer = defines the tensorboard writer
a = A()  # an object of class A defined in A.py

# there is some training stuff going on and I am writing to metrics x, y at iteration i
writer.add_scalar(x, global_step=i)
writer.add_scalar(y, global_step=i)

# I then pass the writer to a member function of class A
# Inside the foo function, I have writer.add_scalar(z, global_step=j)
# j is not same as i
a.foo(writer)
```

When I visualize the plots of `x, y, z` in wandb UI. I have `x` and `y` shown correctly but `z` is not shown with the correct x-axis even though I set x-axis to be `global_step`. 

I think this is because two python files are writing to the same tensorboar writer, i.e., the same writer is passed to and then accessed by different python files? Let me know if any of these is not clear. Many thanks!

**Comment by jingxixu** [comment]
I just figured that it has NOTHING to do with my file structure. This bug will HAPPEN even in an SINGLE file. I have produced a very simple script to reproduce the tensorboard sync issue.

```
from torch.utils.tensorboard import SummaryWriter
import wandb
import numpy as np

wandb.init(project='tensorboard_sync', sync_tensorboard=True, save_code=True)
# this should appear after wandb.init
writer = SummaryWriter(f"{'tensorboard_sync'}")


j = 0
for i in range(1000000):
    if i % 1000 == 0:
        writer.add_scalar('metric_a', i / 100, global_step=j)
        j += 1
    writer.add_scalar('metric_b', np.sin(i), global_step=i)
    writer.add_scalar('metric_c', np.cos(i), global_step=i)
writer.close()
```

Simply run this script you will find that `matric_a` is not correctly displayed with its own customized `global_step`. Please help, I have been stuck in this for a while and it would be awesome if this tensorboard_sync argument can work as expected. Thanks!

**Comment by vanpelt** [comment]
Hey @jingxixu this case is going to be rather tricky for us to handle.  We'll try to take a look this week and see if we can come up with any workarounds, but until then the best solution is to make calls to `wandb.log` without tensorboard in this case.

**Comment by vanpelt** [comment]
Hey @jingxixu the easiest way to fix this is to use a separate writer for the metric that is using a different step.  When using multiple writers it's important to set the root root_logdir for our tensorboard syncing.  The following would work:

```
from torch.utils.tensorboard import SummaryWriter
import wandb
import numpy as np

root_logdir = "tensorboard_sync"
wandb.tensorboard.patch(root_logdir=root_logdir)
wandb.init(project='tensorboard_sync', save_code=True)
# this should appear after wandb.init
train_writer = SummaryWriter(f"{root_logdir}/train")
val_writer = SummaryWriter(f"{root_logdir}/val")

j = 0
for i in range(1000000):
    if i % 1000 == 0:
        val_writer.add_scalar('metric_a', i / 100, global_step=j)
        j += 1
    train_writer.add_scalar('metric_b', np.sin(i), global_step=i)
    train_writer.add_scalar('metric_c', np.cos(i), global_step=i)
train_writer.close()
val_writer.close()
```

Then you can choose "val/global_step" as the x-axis for "metric_a".

**Comment by ajabini** [comment]
> Hey @jingxixu the easiest way to fix this is to use a separate writer for the metric that is using a different step. When using multiple writers it's important to set the root root_logdir for our tensorboard syncing. The following would work:
> 
> ```
> from torch.utils.tensorboard import SummaryWriter
> import wandb
> import numpy as np
> 
> root_logdir = "tensorboard_sync"
> wandb.tensorboard.patch(root_logdir=root_logdir)
> wandb.init(project='tensorboard_sync', save_code=True)
> # this should appear after wandb.init
> train_writer = SummaryWriter(f"{root_logdir}/train")
> val_writer = SummaryWriter(f"{root_logdir}/val")
> 
> j = 0
> for i in range(1000000):
>     if i % 1000 == 0:
>         val_writer.add_scalar('metric_a', i / 100, global_step=j)
>         j += 1
>     train_writer.add_scalar('metric_b', np.sin(i), global_step=i)
>     train_writer.add_scalar('metric_c', np.cos(i), global_step=i)
> train_writer.close()
> val_writer.close()
> ```
> 
> Then you can choose "val/global_step" as the x-axis for "metric_a".

Hi and thanks for this answer. 
I did this and I have to go and manually change the x axis as the plot for metric_a still uses the other step as x-axis. 
I was wondering if there is a way that the plots for metric_a automatically uses j (val_gobal_step)?

**Comment by fdsig** [comment]
Hey @ajabini , 

Thanks so much for messaging with this. I have taken a look at this and I can see that the metric you are looking for is in the val loader will show in the validation (Val) section of your workspace (screenshot below) and shows by default `metric_a`  that you are looking to show on the x-axis. 

![Screenshot 2023-04-28 at 16 45 01](https://user-images.githubusercontent.com/65113682/235193640-3441f042-c450-4e51-a062-d34b396aa825.png)

If for example, you wanted to show `metric_a`, `metric_b`  and `metric_c` side by side then simply move all three default charts (into the same section of your workspace),  `metric_a` from the validation section and metrics  `metric_b`  and `metric_c` from the training section. 

![Screenshot 2023-04-28 at 16 39 30](https://user-images.githubusercontent.com/65113682/235194186-225582ac-7a27-4c2e-86ee-4890d954aad7.png)

[here](https://wandb.ai/demonstrations/tensorboard_sync) is an example open workspace that I made for you to illustrate this. I also appreciate that the 'if i % 1000 == 0:' might make it look like monotonic steps are being logged as metric a and hence be a little confusing <^_^>. Please let me know if you need any further help on this.

**Comment by fdsig** [comment]
Hey @ajabini, wanted to check in here and see if the above worked for you and matches your use case. Look forward to hearing back.

**Comment by fdsig** [comment]
Hey @ajabini as we have not heard back going to go ahead and close this off for you but feel free to reach out if you need any further help or assistance on this.

**Comment by ringohoffman** [comment]
Why don't plots imported from tensorboard just use the same `global_step` that they are logged to tensorboard with?

<img width="493" alt="Screenshot 2023-09-16 at 12 47 12 AM" src="https://github.com/wandb/wandb/assets/27844407/7807a4ca-959d-4e85-9362-5ad179430c17">
<img width="561" alt="Screenshot 2023-09-16 at 12 47 18 AM" src="https://github.com/wandb/wandb/assets/27844407/946bcfc8-5886-4f6e-8f21-d12d2f684c4a">

Ignoring the `global_step` specified to tensorboard and converting it to a wandb-specific global step severely hinders the interpretability of graphs imported from tensorboard.

If I want to log both batch and epoch-level statistics, naturally I will log the batch-level statistics before I can aggregate and log my epoch-level statistics. But since wandb's global_step always increases, the first step of my epoch-level statistics don't start from 1. In the above graph, it starts at 24... and if I change my batch size, this number changes again. The `global_step` that users log to tensorboard with is crucial to interpreting the logged data and is not something that should be discarded.

**Comment by sydholl** [comment]
WandB Internal User commented: 
fdsig commented: 
Hey @ajabini , 

Thanks so much for messaging with this. I have taken a look at this and I can see that the metric you are looking for is in the val loader will show in the validation (Val) section of your workspace (screenshot below) and shows by default `metric_a`  that you are looking to show on the x-axis. 

![Screenshot 2023-04-28 at 16 45 01](https://user-images.githubusercontent.com/65113682/235193640-3441f042-c450-4e51-a062-d34b396aa825.png)

If for example, you wanted to show `metric_a`, `metric_b`  and `metric_c` side by side then simply move all three default charts (into the same section of your workspace),  `metric_a` from the validation section and metrics  `metric_b`  and `metric_c` from the training section. 

![Screenshot 2023-04-28 at 16 39 30](https://user-images.githubusercontent.com/65113682/235194186-225582ac-7a27-4c2e-86ee-4890d954aad7.png)

[here](https://wandb.ai/demonstrations/tensorboard_sync) is an example open workspace that I made for you to illustrate this. I also appreciate that the 'if i % 1000 == 0:' might make it look like monotonic steps are being logged as metric a and hence be a little confusing <^_^>. Please let me know if you need any further help on this.

**Comment by sydholl** [comment]
WandB Internal User commented: 
fdsig commented: 
Hey @ajabini, wanted to check in here and see if the above worked for you and matches your use case. Look forward to hearing back.

**Comment by N-Demir** [comment]
> Why don't plots imported from tensorboard just use the same `global_step` that they are logged to tensorboard with?
> 
> <img alt="Screenshot 2023-09-16 at 12 47 12 AM" width="493" src="https://private-user-images.githubusercontent.com/27844407/268436034-7807a4ca-959d-4e85-9362-5ad179430c17.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA2MDQ5NjYsIm5iZiI6MTc0MDYwNDY2NiwicGF0aCI6Ii8yNzg0NDQwNy8yNjg0MzYwMzQtNzgwN2E0Y2EtOTU5ZC00ZTg1LTkzNjItNWFkMTc5NDMwYzE3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI2VDIxMTc0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQzMzYwMGI4ZTZlNGUzODY5MmIyMzQxOGNkYjM5MzhjN2Y0MjNiN2E5ZmNiNWVhZmM4ZmQ1YzMzYzdhNGYwOWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Gh6_-OWW4FUDBdjtnUQQ9XFrJp1_FOVY8VxjdpE3RCc"> <img alt="Screenshot 2023-09-16 at 12 47 18 AM" width="561" src="https://private-user-images.githubusercontent.com/27844407/268436017-946bcfc8-5886-4f6e-8f21-d12d2f684c4a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA2MDQ5NjYsIm5iZiI6MTc0MDYwNDY2NiwicGF0aCI6Ii8yNzg0NDQwNy8yNjg0MzYwMTctOTQ2YmNmYzgtNTg4Ni00ZjZlLThmMjEtZDEyZDJmNjg0YzRhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI2VDIxMTc0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE1ODE2M2U5MTllNDg4YzgwMDZhOTYxNzUwMWYyNWViMDUwMGZmZTZkYzc4MGI1OWUwYjIyYzkxZjY2ZjAwNWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.fOiYqWqd7Z7mWap30dCuJXdRD1qHED7oLwX7xrZ2odw">
> Ignoring the `global_step` specified to tensorboard and converting it to a wandb-specific global step severely hinders the interpretability of graphs imported from tensorboard.
> 
> If I want to log both batch and epoch-level statistics, naturally I will log the batch-level statistics before I can aggregate and log my epoch-level statistics. But since wandb's global_step always increases, the first step of my epoch-level statistics don't start from 1. In the above graph, it starts at 24... and if I change my batch size, this number changes again. The `global_step` that users log to tensorboard with is crucial to interpreting the logged data and is not something that should be discarded.

I'm having the same issue. Any resolution?

**Comment by ringohoffman** [comment]
@N-Demir Actually, I do have a solution. You can configure a default x-axis for your `wandb` plots using [`define_metric`](https://docs.wandb.ai/ref/python/run/#define_metric). The `tensorboard` step is called `global_step` by `wandb`. You can make it the default x-axis by doing this:

```python
wandb_run.define_metric("*", step_metric="global_step", step_sync=True)
```
